pallets/relay-storage-roots/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+//! roots are removed first, there is an additional `RelayStorageRootKeys` storage item that stores
+//! a sorted list of all the keys. This is needed because it is not possible to iterate over a
+//! mapping in order (unless if using an `Identity` hash). The `MaxStorageRoots` limit applies to
+//! the number of items, not to their age.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use crate::weights::WeightInfo;
+use cumulus_pallet_parachain_system::RelaychainStateProvider;
+use cumulus_primitives_core::relay_chain::BlockNumber as RelayBlockNumber;
+use frame_support::pallet;
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use sp_core::Get;
+use sp_core::H256;
+use sp_std::collections::vec_deque::VecDeque;
+
+#[cfg(any(test, feature = "runtime-benchmarks"))]
+mod benchmarks;
+pub mod weights;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RelaychainStateProvider: RelaychainStateProvider;
+		#[pallet::constant]
+		type MaxStorageRoots: Get<u32>;
+		/// Weight info
+		type WeightInfo: WeightInfo;
+	}
+
+	/// Map of relay block number to relay storage root
+	#[pallet::storage]
+	pub type RelayStorageRoot<T: Config> =
+		StorageMap<_, Twox64Concat, RelayBlockNumber, H256, OptionQuery>;
+
+	/// List of all the keys in `RelayStorageRoot`.
+	/// Used to remove the oldest key without having to iterate over all of them.
+	#[pallet::storage]
+	pub type RelayStorageRootKeys<T: Config> =
+		StorageValue<_, BoundedVec<RelayBlockNumber, T::MaxStorageRoots>, ValueQuery>;
+
+	impl<T: Config> Pallet<T> {
+		/// Populates `RelayStorageRoot` using `RelaychainStateProvider`.
+		pub fn set_relay_storage_root() {
+			let relay_state = T::RelaychainStateProvider::current_relay_chain_state();
+
+			// If this relay block number has already been stored, skip it.
+			if <RelayStorageRoot<T>>::contains_key(relay_state.number) {
+				return;
+			}
+
+			<RelayStorageRoot<T>>::insert(relay_state.number, relay_state.state_root);
+			let mut keys: VecDeque<_> = <RelayStorageRootKeys<T>>::get().into_inner().into();
+			keys.push_back(relay_state.number);
+			// Delete the oldest stored root if the total number is greater than MaxStorageRoots
+			if u32::try_from(keys.len()).unwrap() > T::MaxStorageRoots::get() {
+				let first_key = keys.pop_front().unwrap();
+				<RelayStorageRoot<T>>::remove(first_key);
+			}
+
+			// If `MaxStorageRoots` has decreased, we need to delete more than one storage root.
+			// But that would make this function have an unbounded amount of writes. So instead, we
+			// will only delete the first one, and leak the rest.
+			let keys = BoundedVec::truncate_from(keys.into());
```

Comment:
To avoid that, please add a check here and produce a log at ERROR level if keys.len() > MaxStorageRoots::len()

Reply:
I noticed, added a comment that says that reducing `MaxStorageRoots` needs a migration. It turns out that if the BoundedVec in storage has more than `MaxStorageRoots` items, this is a decode error and the code continues with an empty `BoundedVec` instead. Fortunately the pallet can keep working after such an error, although the storage map will need manual cleanup.

Reply:
In that case can you log a specific ERROR if the BoundedVec is empty? I want to make sure we have a way to catch up these kind of errors in the future

Reply:
We should already see an error log like this, at least in tests:

```
Jan 15 15:02:31.352 ERROR runtime::storage: Corrupted state at `[39, 146, 80, 38, 154, 198, 247, 185, 41, 171, 15, 234, 212, 170, 115, 216, 21, 43, 230, 113, 139, 98, 123, 15, 236, 68, 237, 150, 128, 118, 171, 178]: Error { cause: None, desc: "BoundedVec exceeds its limit" }`    
```

Reply:
Yes but this generic error is not very explicit, please think about the guys that will debug, can you please add a log? I don't see any downside of doing that

Reply:
I don't know how to distinguish corrupted state vs empty state, if we simply check for an empty vec we will see that log the first time the pallet is added, and also in all the tests...

Reply:
Managed to do it using

```rust
if keys.is_empty() && <RelayStorageRootKeys<T>>::exists() {
```

Also tested that it doesn't result in an extra storage read, because the value is cached I guess